ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java

```diff
@@ -490,6 +500,61 @@ public void helixInitiatedLivenessChangeTest() throws Exception {
     assertStateEquivalency(expectedDownInstances, expectedUpInstances);
   }
 
+  /**
+   * Test helix initiated ideal state change can be correctly handled by {@link HelixClusterManager}.
+   * @throws Exception
+   */
+  @Test
+  public void helixInitiatedIdealStateChangeTest() throws Exception {
+    assumeTrue(!useComposite && !overrideEnabled);
+    verifyInitialClusterChanges((HelixClusterManager) clusterManager, helixCluster, dcs);
+    String resourceName = "newResource";
+    String partitionName = String.valueOf(helixCluster.getAllPartitions().size());
+    IdealState idealState = new IdealState(resourceName);
+    idealState.setPreferenceList(partitionName, new ArrayList<>());
+    helixCluster.addNewResource(resourceName, idealState, localDc);
+    verifyInitialClusterChanges((HelixClusterManager) clusterManager, helixCluster, new String[]{localDc});
+    // localDc should have one more partition compared with remoteDc
+    Map<String, ConcurrentHashMap<String, String>> partitionToResource =
+        ((HelixClusterManager) clusterManager).getPartitionToResourceMap();
+    assertEquals("localDc should have one more partition", partitionToResource.get(localDc).size(),
+        partitionToResource.get(remoteDc).size() + 1);
+    assertTrue(partitionToResource.get(localDc).containsKey(partitionName) && !partitionToResource.get(remoteDc)
+        .containsKey(partitionName));
+  }
+
+  /**
+   * Test that routing table change reflects correct state of each replica and {@link HelixClusterManager} is able to get
+   * replica in required state.
+   */
+  @Test
+  public void routingTableProviderChangeTest() {
```

Comment:
I can't tell where the routing table is being changed here, and how the change listener is triggered.

Reply:
Sorry, I didn't mention in comment. `onRoutingTableChange` is called during the HelixClusterManager instantiation. There is an initial notification from Helix that invokes routing table change.  I will try to add test case that verifies routing table change after initialization.

Reply:
new test case added